packages/server/src/clients/api.ts

```diff
@@ -0,0 +1,103 @@
+import { AdminApiManager } from '../base/api-manager'
+import { ClientTokenService } from '../client-tokens/service'
+import { ProviderService } from '../providers/service'
+import { Provider } from '../providers/types'
+import { Schema } from './schema'
+import { ClientService } from './service'
+import { Client } from './types'
+
+export class ClientApi {
+  constructor(
+    private distributed: DistributedService,
+    private providers: ProviderService,
+    private clients: ClientService,
+    private clientTokens: ClientTokenService
+  ) {}
+
+  setup(router: AdminApiManager) {
+    router.post('/admin/clients', Schema.Api.Create, this.create.bind(this))
+    router.post('/admin/clients/sync', Schema.Api.Sync, this.sync.bind(this))
+  }
+
+  async create(req: Request, res: Response) {
+    const clientId = uuidv4()
+
+    const provider = await this.providers.create(clientId, false)
+    const client = await this.clients.create(provider.id, clientId)
+
+    const rawToken = await this.clientTokens.generateToken()
+    const clientToken = await this.clientTokens.create(client.id, rawToken, undefined)
+
+    res.status(201).send({ id: client.id, token: `${clientToken.id}.${rawToken}` })
+  }
+
+  async sync(req: Request, res: Response) {
+    this.distributed.using(`lock_dyn_client_sync::${req.body.name}`, async () => {
+      await this.syncClient(req, res)
+    })
+  }
+
+  async syncClient(req: Request, res: Response) {
+    const sync = { id: req.body.id, token: req.body.token, name: req.body.name } as {
+      id?: uuid
+      token?: string
+      name: string
+    }
+
+    let client: Client | undefined = undefined
+    let token: string | undefined = undefined
+    let provider: Provider | undefined = undefined
+
+    if (sync.id) {
+      client = await this.clients.fetchById(sync.id)
+      if (client) {
+        token = sync.token
+      }
+    }
+
+    if (!client) {
+      const exisingProvider = await this.providers.fetchByName(sync.name)
+      if (exisingProvider) {
+        const existingClient = await this.clients.fetchByProviderId(exisingProvider.id)
+        if (existingClient) {
+          client = await this.clients.getById(existingClient.id)
+        }
+      }
+    }
+
+    if (!client) {
+      provider = await this.providers.create(sync.name, false)
+      client = await this.clients.create(provider.id)
+    } else {
+      provider = await this.providers.fetchById(client.providerId)
+
+      if (!provider) {
+        provider = await this.providers.create(sync.name, false)
+
+        await this.clients.updateProvider(client.id, provider.id)
+        client = await this.clients.getById(client.id)
+      }
+    }
+
+    if (!token || !(await this.clientTokens.verifyToken(client.id, token))) {
+      const rawToken = await this.clientTokens.generateToken()
+      const clientToken = await this.clientTokens.create(client.id, rawToken, undefined)
+      token = `${clientToken.id}.${rawToken}`
```

Comment:
We should really consider putting this into some util function!

Reply:
For a future PR